File: frontend/src/utils/pendingIssues.ts
import type { GameDetail, GameListItem } from '@/services/types'

export type PendingIssueKey =
  | 'missing-assets'
  | 'missing-wiki'
  | 'missing-files'
  | 'missing-metadata'

export type PendingIssueDetailKey =
  | 'missing-cover'
  | 'missing-banner'
  | 'missing-screenshots'
  | 'missing-wiki-content'
  | 'missing-files-list'
  | 'missing-developer'
  | 'missing-publisher'
  | 'missing-platform'
  | 'missing-summary'

interface PendingIssueDefinition {
  key: PendingIssueKey
  label: string
  description: string
}

interface PendingIssueDetailDefinition {
  key: PendingIssueDetailKey
  label: string
  group: PendingIssueKey
}

export interface PendingIssueEvaluation {
  groups: PendingIssueKey[]
  details: PendingIssueDetailKey[]
  ignoredDetails: PendingIssueDetailKey[]
}

export const pendingIssueDefinitions: PendingIssueDefinition[] = [
  {
    key: 'missing-assets',
    label: '缺少图片',
    description: '封面、横幅或截图未补齐',
  },
  {
    key: 'missing-wiki',
    label: '缺少 Wiki',
    description: '还没有游戏介绍内容',
  },
  {
    key: 'missing-files',
    label: '缺少文件',
    description: '还没有可下载文件条目',
  },
  {
    key: 'missing-metadata',
    label: '基础信息不完整',
    description: '开发商、发行商、平台或简介缺失',
  },
]

export const pendingIssueDetailDefinitions: PendingIssueDetailDefinition[] = [
  {
    key: 'missing-cover',
    label: '缺封面',
    group: 'missing-assets',
  },
  {
    key: 'missing-banner',
    label: '缺横幅',
    group: 'missing-assets',
  },
  {
    key: 'missing-screenshots',
    label: '缺截图',
    group: 'missing-assets',
  },
  {
    key: 'missing-wiki-content',
    label: '缺 Wiki 内容',
    group: 'missing-wiki',
  },
  {
    key: 'missing-files-list',
    label: '缺下载文件',
    group: 'missing-files',
  },
  {
    key: 'missing-developer',
    label: '缺开发商',
    group: 'missing-metadata',
  },
  {
    key: 'missing-publisher',
    label: '缺发行商',
    group: 'missing-metadata',
  },
  {
    key: 'missing-platform',
    label: '缺平台',
    group: 'missing-metadata',
  },
  {
    key: 'missing-summary',
    label: '缺简介',
    group: 'missing-metadata',
  },
]

const pendingIssueDetailGroupMap = pendingIssueDetailDefinitions.reduce<Record<PendingIssueDetailKey, PendingIssueKey>>(
  (acc, item) => {
    acc[item.key] = item.group
    return acc
  },
  {} as Record<PendingIssueDetailKey, PendingIssueKey>,
)

export function getPendingIssueLabel(key?: string | null) {
  return pendingIssueDefinitions.find((item) => item.key === key)?.label || '待处理'
}

export function getPendingIssueDetailLabel(key?: string | null) {
  return pendingIssueDetailDefinitions.find((item) => item.key === key)?.label || '待补充'
}

type PendingGame = GameListItem | GameDetail

const hasDetailPayload = (game: PendingGame): game is GameDetail => {
  return 'screenshots' in game
}

const hasListCounts = (game: PendingGame): game is GameListItem => {
  return 'screenshot_count' in game
}

function hasMeaningfulWikiContent(content?: string | null, html?: string | null) {
  if (content?.trim()) {
    return true
  }

  if (!html?.trim()) {
    return false
  }

  const plainText = html
    .replace(/<style[\s\S]*?<\/style>/gi, ' ')
    .replace(/<script[\s\S]*?<\/script>/gi, ' ')
    .replace(/<[^>]+>/g, ' ')
    .replace(/&nbsp;/gi, ' ')
    .trim()

  return plainText.length > 0
}

export function evaluatePendingIssues(game: PendingGame, ignoredDetails: PendingIssueDetailKey[] = []): PendingIssueEvaluation {
  const details: PendingIssueDetailKey[] = []
  const ignoredSet = new Set(ignoredDetails)

  const hasCover = !!game.cover_image
  const hasBanner = !!game.banner_image
  const hasScreenshots =
    (hasListCounts(game) && game.screenshot_count > 0) ||
    (hasDetailPayload(game) && game.screenshots.length > 0) ||
    (hasListCounts(game) && !!game.primary_screenshot)
  if (!hasCover) {
    details.push('missing-cover')
  }
  if (!hasBanner) {
    details.push('missing-banner')
  }
  if (!hasScreenshots) {
    details.push('missing-screenshots')
  }
  const hasWiki = hasMeaningfulWikiContent(game.wiki_content, game.wiki_content_html)
  if (!hasWiki) {
    details.push('missing-wiki-content')
  }

  const hasFiles =
    (hasDetailPayload(game) && game.files.length > 0) ||
    (hasListCounts(game) && game.file_count > 0)
  if (!hasFiles) {
    details.push('missing-files-list')
  }

  const hasDeveloper =
    (hasDetailPayload(game) && game.developers.length > 0) ||
    (hasListCounts(game) && game.developer_count > 0)
  const hasPublisher =
    (hasDetailPayload(game) && game.publishers.length > 0) ||
    (hasListCounts(game) && game.publisher_count > 0)
  const hasPlatform =
    (hasDetailPayload(game) && game.platforms.length > 0) ||
    (hasListCounts(game) && game.platform_count > 0)
  const hasSummary = !!game.summary?.trim()
  if (!hasDeveloper) {
    details.push('missing-developer')
  }
  if (!hasPublisher) {
    details.push('missing-publisher')
  }
  if (!hasPlatform) {
    details.push('missing-platform')
  }
  if (!hasSummary) {
    details.push('missing-summary')
  }

  const visibleDetails = details.filter((detail) => !ignoredSet.has(detail))
  const groups = Array.from(new Set(
    visibleDetails
      .map((detail) => pendingIssueDetailGroupMap[detail])
      .filter((group): group is PendingIssueKey => Boolean(group)),
  ))

  return {
    groups,
    details: visibleDetails,
    ignoredDetails: details.filter((detail) => ignoredSet.has(detail)),
  }
}

export function isSeverePendingEvaluation(evaluation: PendingIssueEvaluation) {
  return (
    evaluation.details.length >= 3 ||
    evaluation.groups.includes('missing-files') ||
    (evaluation.groups.includes('missing-assets') && evaluation.groups.includes('missing-wiki'))
  )
}

export function getPendingIssues(game: PendingGame, ignoredDetails: PendingIssueDetailKey[] = []): PendingIssueKey[] {
  return evaluatePendingIssues(game, ignoredDetails).groups
}

export function getPendingIssueDetails(game: PendingGame, ignoredDetails: PendingIssueDetailKey[] = []): PendingIssueDetailKey[] {
  return evaluatePendingIssues(game, ignoredDetails).details
}

export function getIgnoredPendingIssueDetails(game: PendingGame, ignoredDetails: PendingIssueDetailKey[] = []) {
  return evaluatePendingIssues(game, ignoredDetails).ignoredDetails
}

export function isSeverePendingGame(game: PendingGame, ignoredDetails: PendingIssueDetailKey[] = []) {
  return isSeverePendingEvaluation(evaluatePendingIssues(game, ignoredDetails))
}

export function matchesPendingIssueEvaluation(evaluation: PendingIssueEvaluation, key?: string | null) {
  if (!key) return true
  return evaluation.groups.includes(key as PendingIssueKey)
}

export function matchesPendingIssue(game: PendingGame, key?: string | null) {
  return matchesPendingIssueEvaluation(evaluatePendingIssues(game), key)
}

File: backend/internal/domain/pending_issues.go
package domain

type PendingIssueKey string
type PendingIssueDetailKey string

const (
	PendingIssueMissingAssets   PendingIssueKey = "missing-assets"
	PendingIssueMissingWiki     PendingIssueKey = "missing-wiki"
	PendingIssueMissingFiles    PendingIssueKey = "missing-files"
	PendingIssueMissingMetadata PendingIssueKey = "missing-metadata"
)

const (
	PendingIssueDetailMissingCover       PendingIssueDetailKey = "missing-cover"
	PendingIssueDetailMissingBanner      PendingIssueDetailKey = "missing-banner"
	PendingIssueDetailMissingScreenshots PendingIssueDetailKey = "missing-screenshots"
	PendingIssueDetailMissingWikiContent PendingIssueDetailKey = "missing-wiki-content"
	PendingIssueDetailMissingFilesList   PendingIssueDetailKey = "missing-files-list"
	PendingIssueDetailMissingDeveloper   PendingIssueDetailKey = "missing-developer"
	PendingIssueDetailMissingPublisher   PendingIssueDetailKey = "missing-publisher"
	PendingIssueDetailMissingPlatform    PendingIssueDetailKey = "missing-platform"
	PendingIssueDetailMissingSummary     PendingIssueDetailKey = "missing-summary"
)

type PendingIssueDefinition struct {
	Key         PendingIssueKey
	Label       string
	Description string
}

type PendingIssueDetailDefinition struct {
	Key   PendingIssueDetailKey
	Label string
	Group PendingIssueKey
}

type PendingIssueDetailState struct {
	Key     PendingIssueDetailKey
	Group   PendingIssueKey
	Ignored bool
	Reason  *string
}

type PendingIssueEvaluation struct {
	Groups  []PendingIssueKey
	Details []PendingIssueDetailState
	Severe  bool
}

type PendingIssueCatalog struct {
	Groups  []PendingIssueDefinition
	Details []PendingIssueDetailDefinition
}

type PendingIssueCountSummary struct {
	Groups       map[PendingIssueKey]int
	IgnoredTotal int
}

type PendingIssueSeverityPolicy struct {
	MinVisibleDetails int
	SevereIfAnyGroup  []PendingIssueKey
	SevereIfAllGroups [][]PendingIssueKey
}

var pendingIssueDefinitions = []PendingIssueDefinition{
	{Key: PendingIssueMissingAssets, Label: "缺少图片", Description: "封面、横幅或截图未补齐"},
	{Key: PendingIssueMissingWiki, Label: "缺少 Wiki", Description: "还没有游戏介绍内容"},
	{Key: PendingIssueMissingFiles, Label: "缺少文件", Description: "还没有可下载文件条目"},
	{Key: PendingIssueMissingMetadata, Label: "基础信息不完整", Description: "开发商、发行商、平台或简介缺失"},
}

var pendingIssueDetailDefinitions = []PendingIssueDetailDefinition{
	{Key: PendingIssueDetailMissingCover, Label: "缺封面", Group: PendingIssueMissingAssets},
	{Key: PendingIssueDetailMissingBanner, Label: "缺横幅", Group: PendingIssueMissingAssets},
	{Key: PendingIssueDetailMissingScreenshots, Label: "缺截图", Group: PendingIssueMissingAssets},
	{Key: PendingIssueDetailMissingWikiContent, Label: "缺 Wiki 内容", Group: PendingIssueMissingWiki},
	{Key: PendingIssueDetailMissingFilesList, Label: "缺下载文件", Group: PendingIssueMissingFiles},
	{Key: PendingIssueDetailMissingDeveloper, Label: "缺开发商", Group: PendingIssueMissingMetadata},
	{Key: PendingIssueDetailMissingPublisher, Label: "缺发行商", Group: PendingIssueMissingMetadata},
	{Key: PendingIssueDetailMissingPlatform, Label: "缺平台", Group: PendingIssueMissingMetadata},
	{Key: PendingIssueDetailMissingSummary, Label: "缺简介", Group: PendingIssueMissingMetadata},
}

var pendingIssueSeverityPolicy = PendingIssueSeverityPolicy{
	MinVisibleDetails: 3,
	SevereIfAnyGroup: []PendingIssueKey{
		PendingIssueMissingFiles,
	},
	SevereIfAllGroups: [][]PendingIssueKey{
		{PendingIssueMissingAssets, PendingIssueMissingWiki},
	},
}

var pendingIssueDefinitionMap = func() map[PendingIssueKey]PendingIssueDefinition {
	items := make(map[PendingIssueKey]PendingIssueDefinition, len(pendingIssueDefinitions))
	for _, item := range pendingIssueDefinitions {
		items[item.Key] = item
	}
	return items
}()

var pendingIssueDetailDefinitionMap = func() map[PendingIssueDetailKey]PendingIssueDetailDefinition {
	items := make(map[PendingIssueDetailKey]PendingIssueDetailDefinition, len(pendingIssueDetailDefinitions))
	for _, item := range pendingIssueDetailDefinitions {
		items[item.Key] = item
	}
	return items
}()

func PendingIssueCatalogDefinitions() PendingIssueCatalog {
	return PendingIssueCatalog{
		Groups:  append([]PendingIssueDefinition(nil), pendingIssueDefinitions...),
		Details: append([]PendingIssueDetailDefinition(nil), pendingIssueDetailDefinitions...),
	}
}

func PendingIssueGroupDefinitions() []PendingIssueDefinition {
	return append([]PendingIssueDefinition(nil), pendingIssueDefinitions...)
}

func PendingIssueDetailDefinitions() []PendingIssueDetailDefinition {
	return append([]PendingIssueDetailDefinition(nil), pendingIssueDetailDefinitions...)
}

func PendingIssueSeverityRules() PendingIssueSeverityPolicy {
	policy := PendingIssueSeverityPolicy{
		MinVisibleDetails: pendingIssueSeverityPolicy.MinVisibleDetails,
		SevereIfAnyGroup:  append([]PendingIssueKey(nil), pendingIssueSeverityPolicy.SevereIfAnyGroup...),
	}
	if len(pendingIssueSeverityPolicy.SevereIfAllGroups) == 0 {
		return policy
	}
	policy.SevereIfAllGroups = make([][]PendingIssueKey, 0, len(pendingIssueSeverityPolicy.SevereIfAllGroups))
	for _, groupSet := range pendingIssueSeverityPolicy.SevereIfAllGroups {
		policy.SevereIfAllGroups = append(policy.SevereIfAllGroups, append([]PendingIssueKey(nil), groupSet...))
	}
	return policy
}

func IsAllowedPendingIssueFilter(value string) bool {
	if value == "" {
		return false
	}
	if _, ok := pendingIssueDefinitionMap[PendingIssueKey(value)]; ok {
		return true
	}
	_, ok := pendingIssueDetailDefinitionMap[PendingIssueDetailKey(value)]
	return ok
}

func IsAllowedPendingIssueDetail(value string) bool {
	if value == "" {
		return false
	}
	_, ok := pendingIssueDetailDefinitionMap[PendingIssueDetailKey(value)]
	return ok
}

func PendingIssueDetailDefinitionForKey(key PendingIssueDetailKey) (PendingIssueDetailDefinition, bool) {
	definition, ok := pendingIssueDetailDefinitionMap[key]
	return definition, ok
}

func PendingIssueFilterMatches(filter string, detailKey PendingIssueDetailKey) bool {
	if filter == "" {
		return false
	}
	detail, ok := pendingIssueDetailDefinitionMap[detailKey]
	if !ok {
		return false
	}
	return filter == string(detailKey) || filter == string(detail.Group)
}

func IsPendingIssueSevere(groups []PendingIssueKey, visibleDetailCount int) bool {
	if visibleDetailCount >= pendingIssueSeverityPolicy.MinVisibleDetails {
		return true
	}

	visibleGroups := make(map[PendingIssueKey]struct{}, len(groups))
	for _, group := range groups {
		visibleGroups[group] = struct{}{}
	}

	for _, group := range pendingIssueSeverityPolicy.SevereIfAnyGroup {
		if _, ok := visibleGroups[group]; ok {
			return true
		}
	}

	for _, requiredGroups := range pendingIssueSeverityPolicy.SevereIfAllGroups {
		matched := true
		for _, group := range requiredGroups {
			if _, ok := visibleGroups[group]; !ok {
				matched = false
				break
			}
		}
		if matched {
			return true
		}
	}

	return false
}

type pendingIssueGameFields struct {
	Summary           *string
	CoverImage        *string
	BannerImage       *string
	WikiContent       *string
	PrimaryScreenshot *string
	ScreenshotCount   int64
	FileCount         int64
	DeveloperCount    int64
	PublisherCount    int64
	PlatformCount     int64
}

func EvaluatePendingIssues(game Game, ignoredReasons map[PendingIssueDetailKey]*string) PendingIssueEvaluation {
	return evaluatePendingIssues(pendingIssueGameFields{
		Summary:           game.Summary,
		CoverImage:        game.CoverImage,
		BannerImage:       game.BannerImage,
		WikiContent:       game.WikiContent,
		PrimaryScreenshot: game.PrimaryScreenshot,
		ScreenshotCount:   game.ScreenshotCount,
		FileCount:         game.FileCount,
		DeveloperCount:    game.DeveloperCount,
		PublisherCount:    game.PublisherCount,
		PlatformCount:     game.PlatformCount,
	}, ignoredReasons)
}

func EvaluatePendingIssuesForListItem(game GameListItem, ignoredReasons map[PendingIssueDetailKey]*string) PendingIssueEvaluation {
	return evaluatePendingIssues(pendingIssueGameFields{
		Summary:           game.Summary,
		CoverImage:        game.CoverImage,
		BannerImage:       game.BannerImage,
		WikiContent:       game.WikiContent,
		PrimaryScreenshot: game.PrimaryScreenshot,
		ScreenshotCount:   game.ScreenshotCount,
		FileCount:         game.FileCount,
		DeveloperCount:    game.DeveloperCount,
		PublisherCount:    game.PublisherCount,
		PlatformCount:     game.PlatformCount,
	}, ignoredReasons)
}

func evaluatePendingIssues(game pendingIssueGameFields, ignoredReasons map[PendingIssueDetailKey]*string) PendingIssueEvaluation {
	details := make([]PendingIssueDetailState, 0, len(pendingIssueDetailDefinitions))
	visibleGroups := make(map[PendingIssueKey]struct{}, len(pendingIssueDefinitions))
	visibleDetailCount := 0

	appendDetail := func(key PendingIssueDetailKey) {
		definition, ok := pendingIssueDetailDefinitionMap[key]
		if !ok {
			return
		}
		reason, ignored := ignoredReasons[key]
		details = append(details, PendingIssueDetailState{
			Key:     key,
			Group:   definition.Group,
			Ignored: ignored,
			Reason:  reason,
		})
		if !ignored {
			visibleGroups[definition.Group] = struct{}{}
			visibleDetailCount += 1
		}
	}

	hasMeaningfulWikiContent := game.WikiContent != nil && *game.WikiContent != ""

	if game.CoverImage == nil || *game.CoverImage == "" {
		appendDetail(PendingIssueDetailMissingCover)
	}
	if game.BannerImage == nil || *game.BannerImage == "" {
		appendDetail(PendingIssueDetailMissingBanner)
	}
	if game.ScreenshotCount <= 0 && (game.PrimaryScreenshot == nil || *game.PrimaryScreenshot == "") {
		appendDetail(PendingIssueDetailMissingScreenshots)
	}
	if !hasMeaningfulWikiContent {
		appendDetail(PendingIssueDetailMissingWikiContent)
	}
	if game.FileCount <= 0 {
		appendDetail(PendingIssueDetailMissingFilesList)
	}
	if game.DeveloperCount <= 0 {
		appendDetail(PendingIssueDetailMissingDeveloper)
	}
	if game.PublisherCount <= 0 {
		appendDetail(PendingIssueDetailMissingPublisher)
	}
	if game.PlatformCount <= 0 {
		appendDetail(PendingIssueDetailMissingPlatform)
	}
	if game.Summary == nil || *game.Summary == "" {
		appendDetail(PendingIssueDetailMissingSummary)
	}

	groups := make([]PendingIssueKey, 0, len(pendingIssueDefinitions))
	for _, definition := range pendingIssueDefinitions {
		if _, ok := visibleGroups[definition.Key]; ok {
			groups = append(groups, definition.Key)
		}
	}

	return PendingIssueEvaluation{
		Groups:  groups,
		Details: details,
		Severe:  IsPendingIssueSevere(groups, visibleDetailCount),
	}
}
